libp2p-identify/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the `/ipfs/id/1.0.0` protocol. Allows a node A to query another node B which
+//! information B knows about A. Also includes the addresses B is listening on.
+//!
+//! When two nodes connect to each other, the listening half sends a message to the dialing half,
+//! indicating the information, and then the protocol stops.
+
+extern crate bytes;
+extern crate futures;
+extern crate multiaddr;
+extern crate libp2p_peerstore;
+extern crate libp2p_swarm;
+extern crate protobuf;
+extern crate tokio_io;
+
+use bytes::{Bytes, BytesMut};
+use futures::{Future, Stream, Sink};
+use libp2p_swarm::{ConnectionUpgrade, Endpoint};
+use multiaddr::Multiaddr;
+use protobuf::Message as ProtobufMessage;
+use protobuf::core::parse_from_bytes as protobuf_parse_from_bytes;
+use protobuf::repeated::RepeatedField;
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
+use std::iter;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_io::codec::length_delimited;
+
+mod structs_proto;
+
+/// Prototype for an upgrade to the identity protocol.
+#[derive(Debug, Clone)]
+pub struct IdentifyProtocol {
+	pub information: IdentifyInfo,
+}
+
+/// Information sent from the listener to the dialer.
+#[derive(Debug, Clone)]
+pub struct IdentifyInfo {
+	/// Public key of the node.
+	pub public_key: Vec<u8>,
+	/// Version of the "global" protocol, eg. `ipfs/1.0.0`.
+	pub protocol_version: String,
+	/// Name and version. Can be thought as similar to the `User-Agent` header of HTTP.
+	pub agent_version: String,
+	/// Addresses that are listened on.
+	pub listen_addrs: Vec<Multiaddr>,
+	/// Address that the server uses to communicate with the dialer.
+	pub observed_addr: Multiaddr,
+	/// Protocols supported by the remote.
+	pub protocols: Vec<String>,
+}
+
+impl<C> ConnectionUpgrade<C> for IdentifyProtocol
+    where C: AsyncRead + AsyncWrite + 'static
+{
+	type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+	type UpgradeIdentifier = ();
+
+	#[inline]
+	fn protocol_names(&self) -> Self::NamesIter {
+		iter::once((Bytes::from("/ipfs/id/1.0.0"), ()))
+	}
+
+	type Output = Option<IdentifyInfo>;
+	type Future = Box<Future<Item = Self::Output, Error = IoError>>;
+
+	fn upgrade(self, socket: C, _: (), ty: Endpoint) -> Self::Future {
+		// TODO: use jack's varint library instead
+		let socket = length_delimited::Builder::new().length_field_length(1).new_framed(socket);
+
+		match ty {
```

Comment:
It's a little questionable that this does something totally different depending on whether we're a dialer or a listener. Can we do some type-level stuff to say that some upgrades are only valid on dialers or only valid on listeners?

Reply:
That would require changing half of the design of the library I think. I'm not sure it's worth it.

Reply:
Is it common that an upgrade is valid on only a listener or only a dialer? We have at least one example of a case where we have an upgrade that only supports a single half (i.e. this one), it's not inconceivable that there may be more and so it might be worth doing sooner rather than later so this doesn't become technical debt.